### PR TITLE
[otelcol] Preserve internal representation for components' configurations

### PR DESCRIPTION
**Pull request by mx-psi**
#### Description

The custom unmarshalling for components copied a map without preserving the internal representation. This led to the issues mentioned on #10552 not being fully fixed (they were only fixed if they happened in the `service::telemetry` section for example).

#### Link to tracking issue
Fixes issues mentioned on #10552

#### Testing

This adds one unit test at the `otelcol` level. Since we didn't catch this with our current `confmap/internal/e2e` tests, we likely also want to refactor those.

### CODE REVIEW

**Review by dmitryax — APPROVED** [review]

File: otelcol/unmarshal_dry_run_test.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  nop:
+
+service:
+  pipelines:
+    traces:
+      receivers: [generic]
+      exporters: [nop]
+`
+
+func TestDryRunWithExpandedValues(t *testing.T) {
```
Comment (crobert-1) [inline otelcol/unmarshal_dry_run_test.go:60]:
```suggestion
func TestDryRunWithExpandedValues(_ *testing.T) {
```

```
Error: unmarshal_dry_run_test.go:25:19: unused-parameter: parameter 't' seems to be unused, consider removing or renaming it as _ (revive)
func NewFactories(t *testing.T) func() (Factories, error) {
```

File: otelcol/unmarshal_dry_run_test.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+				Factories: NewFactories(t),
+				ConfigProviderSettings: ConfigProviderSettings{
+					ResolverSettings: confmap.ResolverSettings{
+						URIs:          []string{"file:file"},
+						DefaultScheme: "mock",
+						ProviderFactories: []confmap.ProviderFactory{
+							newFakeProvider("mock", func(_ context.Context, uri string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
+								return confmap.NewRetrievedFromYAML([]byte(tt.mockMap[uri[len("mock:"):]]))
+							}),
+							newFakeProvider("file", func(_ context.Context, uri string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
```
Comment (crobert-1) [inline otelcol/unmarshal_dry_run_test.go:96]:
```suggestion
							newFakeProvider("file", func(_ context.Context, _ string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
```
```
Error: unmarshal_dry_run_test.go:96:56: unused-parameter: parameter 'uri' seems to be unused, consider removing or renaming it as _ (revive)
							newFakeProvider("file", func(_ context.Context, uri string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
```

**Review by crobert-1 — COMMENTED** [review]
```
Generated code is out of date, please run "make gogenerate" and commit the changes in this PR.
```
Just comments from failing check and lint.

File: otelcol/unmarshal_dry_run_test.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+var _ component.Config = (*Config)(nil)
+
+type ValidateTestConfig struct {
+	Number int    `mapstructure:"number"`
+	String string `mapstructure:"string"`
+}
+
+var genericType component.Type = component.MustNewType("generic")
+
+func NewFactories(t *testing.T) func() (Factories, error) {
```
Comment (crobert-1) [inline otelcol/unmarshal_dry_run_test.go:25]:
```suggestion
func NewFactories(_ *testing.T) func() (Factories, error) {
```
```
Error: unmarshal_dry_run_test.go:25:19: unused-parameter: parameter 't' seems to be unused, consider removing or renaming it as _ (revive)
func NewFactories(t *testing.T) func() (Factories, error) {
```

**Review by TylerHelmuth — APPROVED** [review]

File: otelcol/unmarshal_dry_run_test.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package otelcol
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component"
```
Comment (crobert-1) [inline otelcol/unmarshal_dry_run_test.go:11]:
```suggestion

	"go.opentelemetry.io/collector/component"
```
Fix lint

**Review by crobert-1 — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Pull request merged**